doc/ROT13.1
'\" t
.\"     Title: rot13
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.78.1 <http://docbook.sf.net/>
.\"      Date: 06/06/2015
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "ROT13" "1" "06/06/2015" "\ \&" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
ROT13 \- Rotate by 13 places encryption
.SH "SYNOPSIS"
.sp
\fBROT13\fR [\fIOPTIONS\fR] [\fIFILE\fR\&...]
.SH "DESCRIPTION"
.sp
The ROT13(1) command encrypt the file and output it as \fIfile\&.out\fR\&. If no filename is provided, it will read from stdin and output to stdout\&. (see \fBEXAMPLES\fR)\&.
.SH "OPTIONS"
.PP
\fB\-r, \-\-reverse\fR
.RS 4
Reverse shift\&. The default shift direction is left shift\&. This option set the program to perform right shift instead\&.
.RE
.PP
\fB\-s, \-\-shift\fR=\fIAMOUNT\fR
.RS 4
Shift by this AMOUNT, the default AMOUNT is 13\&.
.RE
.PP
\fB\-v, \-\-verbose\fR
.RS 4
Verbosely print processing information\&.
.RE
.PP
\fB\-?, \-\-help\fR
.RS 4
Display help list\&.
.RE
.PP
\fB\-V, \-\-version\fR
.RS 4
Print program version\&.
.RE
.SH "EXAMPLES"
.PP
\fBROT13\fR
.RS 4
Left shift of 13, reading from stdin
.RE
.PP
\fBROT13 \-s 10\fR
.RS 4
Left shift of 10
.RE
.PP
\fBROT13 \-r \-s 8\fR
.RS 4
Right shift of 8
.RE
.PP
\fBROT13 Makefile\fR
.RS 4
Encrypt the Makefile, output as Makefile\&.out
.RE
.SH "EXIT STATUS"
.PP
\fB0\fR
.RS 4
Success
.RE
.PP
\fB1\fR
.RS 4
Failure (syntax or usage error; configuration error; document processing failure; unexpected error)\&.
.RE
.SH "BUGS"
.sp
N/A\&.
.SH "AUTHOR"
.sp
Written by Star Poon <oneonestar@gmail\&.com>
.SH "RESOURCES"
.sp
Source code: https://github\&.com/oneonestar/LSP_ROT13
.SH "COPYING"
.sp
Copyright (C) 2015 Star Poon\&. License GPLv3+: GNU GPL version 3 or later http://gnu\&.org/licenses/gpl\&.html\&. This is free software: you are free to change and redistribute it\&. There is NO WARRANTY, to the extent permitted by law\&.


src/ROT13.c
/**
* @file ROT13.c
* @brief Implementation of Rotate by 13 places (ROT-13).
* @author Star Poon <oneonestar@gmail.com>
* @version 1.0
* @copyright 2015
*
* @section LICENSE
* Copyright (C) 2015 Star Poon
*
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
* 
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
* GNU General Public License for more details.
* 
* You should have received a copy of the GNU General Public License
* along with this program. If not, see <http://www.gnu.org/licenses/>.
*/

#include <stdio.h>
#include <stdlib.h>

#include <errno.h>
#include <string.h>
#include <argp.h>

// Print --version
const char *argp_program_version =
"ROT13 1.0.0\n\
Copyright (C) 2015 Star Poon\n\
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.\n\
This is free software: you are free to change and redistribute it.\n\
There is NO WARRANTY, to the extent permitted by law.\n\
\n\
Written by Star Poon.";

// Print --help
const char *argp_program_bug_address =
  "Star Poon <oneonestar@gmail.com>";

// A descriptive string about this program.
static char doc[] =
  "ROT13 -- Rotate by 13 places encryption";

// Optional file name
static char args_doc[] = "[FILE...]";

// The options that are allowed
static struct argp_option options[] = {
        {"shift",   's',    "AMOUNT",      0,  "Shift by this AMOUNT (default 13)" },
        {"reverse", 'r',    0,      0,  "Reverse shift (right shift)" },
        {"verbose", 'v',    0,      0,  "Verbose Mode" },
        { 0 }
};

// Used by main to communicate with parse_opt
struct arguments
{
    int shift, reverse, verbose;
    char **strings;               /* [FILE…] */
};
static struct arguments arguments = {
        .shift = 13
};

/* 
 * ===  FUNCTION  ======================================================================
 *         Name:  mod
 *  Description:  Simple implementation of mod operator
 * =====================================================================================
 */
unsigned mod(int a, unsigned b) {
	return (a >= 0 ? a % b : b - (-a) % b);
}

/**
 * @brief Error handling for strtol(). strtol() is the safe version of atoi().
 */
int parseLong(const char *str, int *ret)
{
    errno = 0;
    char *temp;
    long val = strtol(str, &temp, 0);

    if (temp == str || *temp != '\0' ||
        ((val == LONG_MIN || val == LONG_MAX) && errno == ERANGE))
        return 1;   //Failed to convert to int
    *ret = (int)val;
    return 0;
}

/**
 * @brief Parse a single option.
 */
static error_t
parse_opt (int key, char *arg, struct argp_state *state)
{
    /* Get the input argument from argp_parse, which we
       know is a pointer to our arguments structure. */
    int ret, arg_int;
    struct arguments *arguments = state->input;
    switch (key)
    {
        case 's':
            ret = parseLong(arg, &arg_int);
            if(ret) {
                // Cannot convert arg to number, print error and exit()
                argp_error(state, "%s", strerror(EINVAL));
            }
            arguments->shift = arg_int;
            break;
        case 'r':
            arguments->reverse = 1;
            break;
        case 'v':
            arguments->verbose = 1;
            break;
        case ARGP_KEY_ARG:
            arguments->strings = &state->argv[state->next-1];
            state->next = state->argc;
            break;
        default:
            return ARGP_ERR_UNKNOWN;
    }
    return 0;
}

/**
 * @brief The actual convertion algorithm
 */
int convert_file(FILE *inFile, FILE *outFile)
{
    do {
        // Read a character from stdin
        int c = fgetc(inFile);
        // End of file? Nothing left to read...
        if (feof(inFile)) {
            // Exit loop
            break;
        }

        // Encrypt
        if (isupper(c)) {
            // Shift
            c = arguments.reverse? c+arguments.shift: c-arguments.shift;

            // Range check
            if (c > 'Z') {
                c -= 26;
            }
            else if (c < 'A') {
                c += 26;
            }
        }

        if (islower(c)) {
            // Shift
            c = arguments.reverse? c+arguments.shift: c-arguments.shift;

            // Range check
            if (c > 'z') {
                c -= 26;
            }
            else if (c < 'a') {
                c += 26;
            }
        }
        // Print the character to stdout (default)
        fprintf(outFile, "%c", c);
    } while (1);
    return 0;
}

// Used by argp
static struct argp argp = { options, parse_opt, args_doc, doc };

char c;
int main (int argc, char **argv)
{
    // Calling the argp lib to handle the arguments
    error_t ret = argp_parse(&argp, argc, argv, 0, 0, &arguments);
    if(ret) {
        fprintf(stderr, "%s: Unknown error during parsing argument: %s\n", argv[0], strerror(ret));
        return EXIT_FAILURE;
    }
    if (arguments.verbose) {
        printf("Shift=%d, Reverse=%d\n", arguments.shift, arguments.reverse);
    }

    if (arguments.strings) {
        // Provided File Name -> Convert files
        for (int j = 0; arguments.strings[j]; j++) {
            FILE *inFile, *outFile;
            inFile = fopen (arguments.strings[j],"r");
            if(!inFile) {
                fprintf(stderr, "%s: Failed to open file '%s': %s\n", argv[0], arguments.strings[j], strerror(errno));
                return EXIT_FAILURE;
            }
            char temp_file_name[] = "fileXXXXXX";

            // We use mkstemp() because tmpnam() is not safe.
            int temp_fd = mkstemp(temp_file_name);
            outFile = fdopen(temp_fd, "w");
            if(!outFile) {
                fprintf(stderr, "%s: Failed to create temp file: %s\n", argv[0], strerror(errno));
                return EXIT_FAILURE;
            }
            convert_file(inFile, outFile);

            fclose(inFile);
            fclose(outFile);

            // "filename"+".out" as the output
            if (strlen(arguments.strings[j])+4>=FILENAME_MAX) {
                // if "filename"+".out" is too long, just clean up and terminate the program.
                remove(temp_file_name);
                return EXIT_FAILURE;
            }

            // According to GCC standard, we shouldn't use FILENAME_MAX as the size of an array in which to store a
            // file name.  Use dynamic allocation instead.
            // I doubt that but we just follow it here.
            char *final_output = malloc(strlen(arguments.strings[j])+4);
            // Output as ori_filename.out
            snprintf(final_output, FILENAME_MAX, "%s.out", arguments.strings[j]);
            remove(final_output);
            rename(temp_file_name, final_output);
            if (arguments.verbose) {
                printf("Converting %s\n", arguments.strings[j]);
                printf("Output as %s\n", final_output);
            }
            free(final_output);
        }
    } else {
        // No file name provided -> Convert stdin
        convert_file(stdin, stdout);
    }

    // Exit cleanly
    return EXIT_SUCCESS;

}

